ssl-config/src/main/java/com/palantir/remoting/ssl/SslConfiguration.java

```diff
@@ -20,15 +20,29 @@
 import com.fasterxml.jackson.databind.annotation.JsonSerialize;
 import com.google.common.base.Optional;
 import com.google.common.base.Preconditions;
+import com.google.common.base.Throwables;
+import java.io.IOException;
+import java.io.InputStream;
+import java.nio.file.Files;
 import java.nio.file.Path;
+import java.security.GeneralSecurityException;
+import java.security.Key;
+import java.security.KeyStore;
+import java.security.cert.Certificate;
+import javax.net.ssl.KeyManagerFactory;
+import javax.net.ssl.TrustManagerFactory;
 import org.immutables.value.Value;
 import org.immutables.value.Value.Style.ImplementationVisibility;
 
+/**
+ * Configuration that provides a {@link TrustManagerFactory} based on a trust store and optionally provides
+ * a {@link KeyManagerFactory} based on a key store. The stores can be in JKS or PKCS12 format.
+ */
 @JsonDeserialize(as = ImmutableSslConfiguration.class)
 @JsonSerialize(as = ImmutableSslConfiguration.class)
 @Value.Immutable
 @Value.Style(visibility = ImplementationVisibility.PACKAGE)
-public abstract class SslConfiguration {
+public abstract class SslConfiguration implements SslConfigurationProvider {
```

Comment:
Yeah, I did consider that, and that would probably work fine for this case. I thought it might be a bit awkward since it changes how `type` is interpreted (right now we use it as the store type, but that's not strictly necessary so we could change it to signal "type of config" instead and branch logic based on that), and it's possible that there could be config that either doesn't fit the fields described in this particular object or require more fields.

However, doing this based on type would keep the storage simpler (moves the branching of logic to the class that handles the object instead), and for the current file/directory proposal, the API should work.

I'll push up a PR with this approach to compare.